lib/context_adapter_helper.js

```diff
@@ -573,94 +689,65 @@ function sendThirdPartyRequest(updateContextRequest, callback) {
- * @param {function} callback Callback to be called informing of any errors or returning
- *  the result of the notification
- */
-function notifyOperationInProgress(request, response, callback) {
-  var payload = {
-    contextElements: [
-      {
-        id: request.contextAdapter.operationDescriptor.buttonId,
-        type: caConfig.BUTTON_ENTITY.TYPE,
-        isPattern: false,
-        attributes: [
-          {
-            name: caConfig.BUTTON_ENTITY.EXTERNAL_ID_ATTR_NAME,
-            type: caConfig.ATTRIBUTE_TYPE.STRING,
-            value: getExternalRequestId(response)
-          },
-          {
-            name: caConfig.BUTTON_ENTITY.OPERATION_STATUS_ATTR_NAME,
-            type: caConfig.ATTRIBUTE_TYPE.STRING,
-            value: caConfig.BUTTON_ENTITY.OPERATION_STATUS.IN_PROGRESS
-          }
-        ]
-      }
-    ]
-  };
-  sendUpdateContext(
-    request.headers['fiware-service'],
-    request.headers['fiware-servicepath'],
-    payload,
-    callback
-  );
-}
-
-/**
  * Sends a operation completed notification to the Context Broker
- * @param {object} request The updateContext request received by the Context Adapter
- *   or the update request received from a third party. In the latter case and until
- *   further specified, the request's payload should contain the associated operation descriptor
- * @param {object} response The response received by the third party if the request is
+ * @param {object} request The updateContext request received by the Context Adapter if available
+ * @param {object} data The data received from the Third Party as the operation request result
  *  an updateContext request, null if the request is an update request sent by a third party.
  * @param {function} callback Callback to be called informing of any errors or returning
  *  the result of the notification
  */
-function notifyOperationCompleted(request, response, callback) {
+function notifyOperationCompleted(request, data, callback) {
+  var contextElements = [
+    {
+      id: request ? request.contextAdapter.operationDescriptor.buttonId :
+        data.payload.buttonId,
+      type: caConfig.BUTTON_ENTITY.TYPE,
+      isPattern: false,
+      attributes: [
+        {
+          name: caConfig.BUTTON_ENTITY.OPERATION_STATUS_ATTR_NAME,
+          type: caConfig.ATTRIBUTE_TYPE.STRING,
+          value: caConfig.BUTTON_ENTITY.OPERATION_STATUS.COMPLETED
+        },
+        {
+          name: caConfig.BUTTON_ENTITY.OPERATION_RESULT_ATTR_NAME,
+          type: caConfig.ATTRIBUTE_TYPE.STRING,
+          value: getResult(null, data)
+        }
+      ]
+    }
+  ];
+
+  if (request && request.contextAdapter.operationDescriptor.interactionType !==
+    caConfig.BUTTON_ENTITY.INTERACTION_TYPES.ASYNCHRONOUS) {
+    contextElements[0].attributes.push({
+      name: caConfig.BUTTON_ENTITY.IOTA_OPERATION_RESULT_ATTR_NAME,
+      type: caConfig.ATTRIBUTE_TYPE.STRING,
+      value: getResult(null, data)
```

Comment:
Marcos is right. What he says is you can store the result of the call to `getResult()` in a variable, then passing this variable to both object literals, instead of calling the function twice. It doesn't have anything to do with the push.